Gentoo-bug: https://bugs.gentoo.org/536426
Reported-by: Eric Siegel
commit 021b7978d14799bae779907faf7490cfd21b3f46
Author: Austin Seipp <austin@well-typed.com>
Date:   Sun Jul 20 10:13:15 2014 -0500

    driver: use absolute paths in ld scripts (#7452)
    
    Patch contributed by slowmo.
    
    Signed-off-by: Austin Seipp <austin@well-typed.com>

diff --git a/compiler/main/DriverPipeline.hs b/compiler/main/DriverPipeline.hs
index 11427e2..49126fe 100644
--- a/compiler/main/DriverPipeline.hs
+++ b/compiler/main/DriverPipeline.hs
@@ -2166,7 +2166,9 @@ joinObjectFiles dflags o_files output_fn = do
   if ldIsGnuLd
      then do
           script <- newTempName dflags "ldscript"
-          writeFile script $ "INPUT(" ++ unwords o_files ++ ")"
+          cwd <- getCurrentDirectory
+          let o_files_abs = map (cwd </>) o_files
+          writeFile script $ "INPUT(" ++ unwords o_files_abs ++ ")"
           ld_r [SysTools.FileOption "" script] ccInfo
      else if sLdSupportsFilelist mySettings
      then do
